internal/hcsoci/resources.go

```diff
@@ -34,151 +32,84 @@ type Resources struct {
-	// (bind-)mounts into a WCOW v2 Xenon.
-	vsmbMounts []string
-
-	// pipeMounts is an array of the host-paths of named pipes mounted into a utility
-	// VM.
-	pipeMounts []string
-
-	// plan9Mounts is an array of all the host paths which have been added to
-	// an LCOW utility VM
-	plan9Mounts []*uvm.Plan9Share
-
-	// netNS is the network namespace
-	netNS string
-
-	// networkEndpoints is the list of network endpoints used by the container
-	networkEndpoints []string
-
+	netNS              string
 	// createNetNS indicates if the network namespace has been created
 	createdNetNS bool
-
 	// addedNetNSToVM indicates if the network namespace has been added to the containers utility VM
 	addedNetNSToVM bool
+	// layers is a pointer to a struct of the layers paths of a container
+	layers *ImageLayers
+	// resources is an array of the resources associated with a container
+	resources []ResourceCloser
+}
 
-	// scsiMounts is an array of the vhd's mounted into a utility VM to support
-	// scsi device passthrough.
-	scsiMounts []scsiMount
+// ResourceCloser is a generic interface for the releasing of a resource. If a resource implements
+// this interface(which they all should), freeing of that resource should entail one call to
+// <resourceName>.Release(ctx, all)
+type ResourceCloser interface {
+	Release(context.Context) error
+}
 
-	// vpciDevices is an array of vpci device names
-	vpciDevices []string
+// AutoManagedVHD struct representing a VHD that will be cleaned up automatically.
+type AutoManagedVHD struct {
+	hostPath string
 }
 
-type scsiMount struct {
-	// path is the host path to the vhd that is mounted.
-	path string
-	// autoManage if `true` means that on cleanup, the runtime should
-	// automatically delete this vhd.
-	autoManage bool
+// Release removes the vhd.
+func (vhd *AutoManagedVHD) Release(ctx context.Context) error {
+	if err := os.Remove(vhd.hostPath); err != nil {
+		log.G(ctx).WithField("hostPath", vhd.hostPath).WithError(err).Error("failed to remove automanage-virtual-disk")
+	}
+	return nil
 }
 
-// TODO: Method on the resources?
+// ReleaseResources releases/frees all of the resources associated with a container. This includes
+// Plan9 shares, vsmb mounts, pipe mounts, network endpoints, scsi mounts, vpci devices and layers.
 func ReleaseResources(ctx context.Context, r *Resources, vm *uvm.UtilityVM, all bool) error {
-	if vm != nil && r.addedNetNSToVM {
-		if err := vm.RemoveNetNS(ctx, r.netNS); err != nil {
-			log.G(ctx).Warn(err)
+	if vm != nil {
+		if r.addedNetNSToVM {
+			if err := vm.RemoveNetNS(ctx, r.netNS); err != nil {
+				log.G(ctx).Warn(err)
+			}
+			r.addedNetNSToVM = false
 		}
-		r.addedNetNSToVM = false
 	}
 
-	if r.createdNetNS {
-		for len(r.networkEndpoints) != 0 {
-			endpoint := r.networkEndpoints[len(r.networkEndpoints)-1]
-			err := hns.RemoveNamespaceEndpoint(r.netNS, endpoint)
-			if err != nil {
-				if !os.IsNotExist(err) {
+	for _, resource := range r.resources {
```

Comment:
I think we should enumerate this backwards, so that the most recently created resource is the first to be cleaned up. It means you'll have to add the automanage-vhd resource before the one for the mount using it (so that we unmount the vhd before deleting it), but I think it will be more correct overall. It also means if we introduce more ordering dependent resources in the future, that they will be cleaned up in the correct order.